### DISCUSSION TITLE
authentication access restrictions

### DISCUSSION THREAD

**Discussion by ghost** [zauberzeug/nicegui#1102] [Q&A] [answered]
### Question

I am tying to learn nicegui by building a CRM website. Can any one help me to understand how can restrict access to different pages based on the user name/role. i have started something but it is not working. 

```py
@ui.page('/admin')
def adminPage():
    if not app.storage.user.get('authenticated', False) and app.storage.user.get('username') != 'admin': 
        return RedirectResponse('/login')
    def regUser(userName,Password,Role):
        conn=sqlite3.connect('userlogin.db')
        users=pd.read_sql_query('SELECT * FROM users',conn)
        if userName == '':
            ui.notify('Username is empty')
        elif len(users[users.username.isin([userName])]) > 0:
            ui.notify("Username Exists", closeBtn='OK', type="negative")
        else:
            curr=conn.cursor()
            curr.execute(f'''INSERT INTO users VALUES ("{userName}","{Password}","{Role}","{datetime.datetime.now()}");''')
            conn.commit()
            curr.close()
            ui.notify(f"The user {userName} is added to the users list", closeBtn='OK', type="positive")
            unameReg.value=""
            pword.value=""
            urole.value=""
```
.

**Comment by rodja** [comment]
Have you seen our [auth-example](https://github.com/zauberzeug/nicegui/blob/main/examples/authentication/main.py)?

**Reply by ghost** [reply]
I did start with that. But being novice i did not know how to put in place access restrictions for pages based on user roles.

**Comment by ProbablyBrian** [comment, marked as answer]
Perhaps your initial "if not authenticated and not named admin" check should be an 'or' instead of 'and'?

Having just solved this in my own project, I think you may be better served taking advantage of FastAPI's dependency injection system for this.  I tossed together a quick gist to illustrate.  It's a haphazard combination of NiceGUI's existing authentication example, some of FastAPI docs' examples, and a bit of user role checking layered on top.

* Any user can access the root homepage.
* Authenticated users of any role can access the user portal.  Anonymous users get an exception.
* The Moderator Dashboard does the permissions checking in the page itself and shows how you might change contents based on permissions (Disabling buttons, showing/hiding elements, etc).
* The Admin Center example follows the Standard User example of raising an exception but does so for insufficient roles instead.

https://gist.github.com/ProbablyBrian/5a47e6ee2242eb0edb572d07b54dcd73

This is just a toy example, of course.  

P.S: I'd strongly discourage formatting/concatenating user input directly into SQL queries...

**Reply by ghost** [reply]
Thank you
